lib/pdcleaner/pdcleaner.go

```diff
@@ -0,0 +1,263 @@
+						log.Errorf("cleaning up boost deal %s for piece %s: %s", d.DealUuid.String(), d.ClientDealProposal.Proposal.PieceCID.String(), err.Error())
+					}
+				}
+			}
+		}
+	}
+
+	// Clean up completed Boost deals
+	for _, d := range legacyDeals {
+		// Confirm deal did not reach termination before Publishing. Otherwise, no need to clean up
+		if d.DealID > abi.DealID(0) {
+			// If deal exists online
+			md, ok := deals[strconv.FormatInt(int64(d.DealID), 10)]
+			if ok {
+				// If deal is slashed or end epoch has passed. No other reason for deal to reach termination
+				if md.Proposal.EndEpoch < head.Height() || md.State.SlashEpoch > 0 {
+					err = p.pd.RemoveDealForPiece(p.ctx, d.ClientDealProposal.Proposal.PieceCID, d.ProposalCid.String())
+					if err != nil {
+						// Don't return if cleaning up a deal results in error. Try them all.
+						log.Errorf("cleaning up legacy deal %s for piece %s: %s", d.ProposalCid.String(), d.ClientDealProposal.Proposal.PieceCID.String(), err.Error())
+					}
+				}
+			}
+		}
+	}
+
+	// Clean up direct deals
+	claims, err := p.full.StateGetClaims(p.ctx, p.miner, tskey)
+	if err != nil {
+		return fmt.Errorf("getting claims for the miner %s: %w", p.miner.String(), err)
+	}
+	for _, d := range completeDirectDeals {
+		// AllocationID and ClaimID should match
+		cID := verifregtypes.ClaimId(d.AllocationID)
+		c, ok := claims[cID]
+		if ok {
+			// TODO: Figure out slashing mechanism in Direct Deals and add that condition here
+			if c.TermMax < head.Height() {
+				err = p.pd.RemoveDealForPiece(p.ctx, d.PieceCID, d.ID.String())
+				if err != nil {
+					// Don't return if cleaning up a deal results in error. Try them all.
+					log.Errorf("cleaning up legacy deal %s for piece %s: %s", d.ID.String(), d.PieceCID, err.Error())
+				}
+			}
+		}
+	}
+
+	// Clean up dangling LID deals with no Boost, Direct or Legacy deals attached to them
+	plist, err := p.pd.ListPieces(p.ctx)
+	if err != nil {
+		return fmt.Errorf("getting piece list from LID: %w", err)
+	}
+
+	for _, piece := range plist {
+		pdeals, err := p.pd.GetPieceDeals(p.ctx, piece)
+		if err != nil {
+			return fmt.Errorf("getting piece deals from LID: %w", err)
+		}
+		for _, deal := range pdeals {
+			// Remove only if the miner ID matches to avoid removing for other miners in case of shared LID
+			if deal.MinerAddr == p.miner {
+
+				bd, err := p.dealsDB.ByPieceCID(p.ctx, piece)
+				if err != nil {
+					return err
+				}
+				if len(bd) > 0 {
+					continue
+				}
+
+				ld, err := p.legacyDeals.ByPieceCid(p.ctx, piece)
+				if err != nil {
+					return err
+				}
+				if len(ld) > 0 {
+					continue
+				}
+
+				dd, err := p.directDealsDB.ByPieceCID(p.ctx, piece)
+				if err != nil {
+					return err
+				}
+				if len(dd) > 0 {
+					continue
+				}
+
+				err = p.pd.RemoveDealForPiece(p.ctx, piece, deal.DealUuid)
+				if err != nil {
+					// Don't return if cleaning up a deal results in error. Try them all.
```

Comment:
I am curious what happens to the dangling deals that fail to get removed?
Ditto re legacy deals above.

Reply:
This is to account for any DB errors. We can always retry in the later attempt. it seem inefficient to stop the cleanup for a single failure as keep that data for a little more time won't hurt the SP.

Reply:
Proceeding seems reasonable. Though I am curious where if ever those lingering deals get cleaned up. 

Context: addressing the cleanup retry for lingering deals can certainly be done in a separate PR. Just now I need your help to understand the systems dynamics in such failure cases.

Reply:
Imagine if there is issue with deleting a particular deal, it could be an ActorState or DB timeout or something else, we will not be cleaning up other deals which could have been cleaned without any issue. If such a problem comes very early in the execution then we keep holding the data we don't need. All the failures are logged and user can easily use CLI to try to remove that deal manually. This step can also help us debug the problem. 
The only exception would be we cannot CLI cannot cleanup deals with no corresponding identifier in SQL. We can handle that later if reported. 

I hope this makes sense.